Sources/MiniDOM.swift

```diff
@@ -17,116 +17,349 @@
     var nodeValue: String? {
         return text
     }
 }
 
 // MARK: - Document
 
-public class Document: Node {
+/**
+ The `Document` class represents the entire document. Conceptually, it is the
+ root of the document tree, and provides the primary access to the document's
+ data.
+ */
+public final class Document: ParentNode {
 
     // MARK: Node
 
+    /// The `nodeType` of a `Document` is `.document`.
     public static let nodeType: NodeType = .document
 
+    /// The `nodeName` of a `Document` is `#document`.
     public let nodeName: String = "#document"
 
+    /// The `nodeValue` of a `Document` is `nil`.
     public let nodeValue: String? = nil
 
+    /**
+     The children of the document will contain the document element and any
+     processing instructions or comments that are siblings of the documnent
+     element.
+     */
     public var children: [Node] = []
 
+    /// The `attributes` dictionary of a `Document` is `nil`.
+    public let attributes: [String : String]? = nil
+
     // MARK: Document
 
+    /**
+     This is a convenience attribute that allows direct access to the child node
+     that is the root element of the document.
+     */
     public var documentElement: Element? {
         return first(in: children, ofType: Element.self)
     }
 }
 
 // MARK: - Element
 
-public class Element: Node {
+/**
+ By far the vast majority of objects (apart from text) that authors encounter
+ when traversing a document are Element nodes. Assume the following XML
+ document:
+
+ ```xml
+ <elementExample id="demo">
+   <subelement1/>
+   <subelement2><subsubelement/></subelement2>
+ </elementExample>
+ ```
+
+ When represented using DOM, the top node is an `Element` node for
+ `elementExample`, which contains two child `Element` nodes, one for
+ `subelement1` and one for `subelement2`. `subelement1` contains no child
+ nodes.
+
+ Elements may have attributes associated with them; since the `Element` class
+ implements the `Node` protocol, the `attributes` property of the `Node`
+ protocol may be used to retrieve the set of all attributes for an element.
+ */
+public final class Element: ParentNode {
 
     // MARK: Node
 
+    /// The `nodeType` of an `Element` is `.element`.
     public static let nodeType: NodeType = .element
 
+    /// The `nodeName` of an `Element` is its `tagName`.
     public var nodeName: String {
         return tagName
     }
 
+    /// The `nodeValue` of an `Element` is `nil`.
     public let nodeValue: String? = nil
 
+    /// The children of the element.
     public var children: [Node] = []
 
     // MARK: Element
 
+    /**
+     The name of the element. For example, in:
+
+     ```xml
+     <elementExample id="demo">
+        ...
+     </elementExample>
+     ```
+
+     `tagName` has the value `"elementExample"`.
+     */
     public var tagName: String
 
+    /**
+     If namespace processing is turned on, contains the URI for the current
+     namespace.
+     */
```

Comment:
`Parser` must provide an initializer argument that maps to `shouldProcessNamespaces`.